File: c/quick.c
#include <stdlib.h>

void swap(int *a, int *b);

// quick implements quick sort algorithm. The function takes an array and its length.
void quick(int a[], int n) {
	if(n <= 1) {
		return;
	}
	swap(&a[0], &a[(int)(rand())%n]);
	int k = 0;
	for(int i = 1; i < n; i++) {
		if(a[i] < a[0])
			swap(&a[++k], &a[i]);
	}
	swap(&a[0], &a[k]);
	quick(a, k);
	quick(a+k+1, n-(k+1));
}

File: php/quick.php
<?php
require_once('util.php');

// quick implements quick sort algorithm. The function takes an array.
function quick(&$a) {
	_quick($a, 0, count($a));
}

function _quick(&$a, $l, $n) {
	if($n <= 1)
		return;
	swap($a, $l, $l+rand(0, $n-1));
	$k = $l;
	for($i = $l+1; $i < $l+$n; $i++) {
		if($a[$i] < $a[$l])
			swap($a, ++$k, $i);
	}
	swap($a, $l, $k);
	_quick($a, $l, $k-$l);
	_quick($a, $k+1, $l+$n-$k-1);
}
